1130 update mission infrastructure

misaugstad opened the pull request:
Resolves #1130 
Resolves #646 
Resolves #1095 

It is finally here! The moment you've all been waiting for... More flexible mission code!!! The code base is 2% to 3% smaller, missions are more oriented towards the user instead of our back-end, you can query for the labels placed during a mission, and so much more!

Testing:
1. I would restore a second database dump in your local environment alongside your main one. I would get an older dump (that is smaller). I called my database `sidewalk-small`. Then edit the `conf/application.conf` file line 38 to have the different database name. I would do this because this update will make changes to the database that can't be undone without restoring the entire database from scratch again.
1. Load up the landing page and check that the tables in the database have been updated appropriately:
    1. The `mission_user` table should be gone
    1. The `mission` table should have completely different columns from the old version
    1. There should be a new `mission_type` table containing 4 entries
    1. The `label`, `audit_task_comment`, `audit_task_interaction`, `audit_task_environment`, and `audit_task_incomplete` tables should all have a new `mission_id` column (and all these tables will now be empty).
    1. The `audit_task_comment` table should now have an `audit_task_id` column.
1. Switch back to the `develop` branch and reload the landing page, it should undo all of those database changes listed above (except it will not fill in the tables that were emptied). You can then go back to the `1130-update-mission-infrastructure` branch.
1. Go through the tutorial and a few missions. I've touched pretty much every part of the interface in this PR, so the person testing needs to look really carefully and every single part of the UI to make sure that it makes sense. Here is a list of some things to check, but please do a sanity check on everything you can see.
    1. The "Current Neighborhood" section in the top-right of the audit page needs to update correctly while you audit, and have correct values when you refresh the page partially through a mission and/or street.
    1. The "Current Mission" section right below that should appropriate text at all times, including saying "Complete the onboarding tutorial!" during onboarding and having the correct mission distance at other times. The mission progress bar should increment as expected. It should get to approximately 100% at the end of the mission.
    1. The label count section below that should work as expected.
    1. The mission start modal (where it says what your next mission is and you click "Ok") should have appropriate text for every mission. So the first mission should have text referring to a "first mission", but only on the first mission. The distance that is given for the mission should be the same as it is shown everywhere else in the UI during the mission.
    1. All the numbers on the mission complete modal should look correct (label counts, distance in the mission, total distance audited in the neighborhood, etc).
1. Make sure that everything works as expected when you use the jump button (both in the beginning, middle, and end of a mission).
1. When auditing a new neighborhood, I have the default sequence of missions to be 500ft, 500ft, 1000ft, 1000ft, then all the remaining being 1/4 mile
1. Check that resuming a mission works correctly: partially audit a street, check the mission completion percentage, refresh the page, and your mission completion percentage should remain the same (it used to reset you back to the completion amount from the start of that street you partially did).
1. Check that the retake tutorial button works as expected.
1. You should be able to refresh the page at weird times without anything looking incorrect in the UI or in the `mission` table. Try refreshing the page mid-mission, on the mission start screen, on the mission complete screen, etc. Check for anything that seems out of place if you do that.
1. Complete a mission. Check the `audit_task_interaction` table for an entry with the action `MissionComplete`. It should now have the metadata `missionId`, `missionType`, `distanceMeters`, and `regionId` in the `note` field. It used to have `missionLabel`, `missionDistance`, and `neighborhoodId`.
1. Check that the "skip tutorial" option works correctly when you are shown the tutorial.
1. Check that there are the correct number of completed missions in the `mission` table. There should be a max of 1 incomplete mission per user per neighborhood.
1. Finish auditing an entire neighborhood (I usually do this by finding a really small one and then jumping through the streets, clicking the "I cannot go in the direction you want me to walk" option). The last mission in the neighborhood should be the exact distance left to audit in that neighborhood (although you will probably hit the "neighborhood complete" screen before you reach 100% completion of the mission, but I don't think that is super important). The page should refresh and you should be sent to a new neighborhood.
1. After completing that neighborhood, start to audit an adjacent neighborhood. In the old code, since neighborhoods shared streets, when you went to this second neighborhood you would be greeted by a mission complete modal and such, which didn't make sense. Instead, it should look like you haven't started auditing this neighborhood yet, except that when you finish a mission, you will notice that the total distance audited in that neighborhood should include the shared streets.
1. Test out starting your auditing by picking a neighborhood through one of the choropleths. Make sure the audit page loads correctly and nothing seems out of place. Make sure that you can resume a partially completed mission that way.
1. If you did the above checks as an anon user, do them again as a registered user, or vice versa.
1. Check that the turker-specific things work correctly
    1. Log in as a turker by opening a new incognito/private browser and using the URL `localhost:9000/?referrer=mturk&hitId=h1&workerId=testTurker1&assignmentId=a1`
    1. You should be sent directly to the audit page. Make sure that the "skip tutorial" is _not_ an option.
    1. Whenever you start a new mission you should also see "Reward on satisfactory completion: $X.XX" on the mission start modal. Make sure that the dollar amount is equal to the distance they are auditing in miles times $4.17 (which would be $0.39 for that first 500 ft mission, for example).
    1. You should see that same mission reward amount in the "Current Mission" section of the UI while the user is auditing, and the "Total Reward Earned" right below that number should also be correct (note that a turker is paid $0.43 for the tutorial). This amount should increase by the correct amount after every mission.
    1. When you finish a mission, the same reward amount for completing a mission should be shown each time.
    1. When you finish your first non-tutorial mission, you should see a confirmation code appear above the continue button. After clicking continue, that button should remain reachable from a button on the left above the sound button.
    1. After the 2nd mission, you should get a short survey that pops up.
    1. Try doing things to mess with the reward amounts by refreshing the page at weird times, entering the audit page through a choropleth, etc etc.
    1. Go to the user dashboard, and you should see the same "Total reward earned" amount, and it should be correct and the same as is shown on the audit page.

jonfroehlich commented:
Woohoooo!

manaswis commented:
Exciting!!!

aileenzeng commented:
Starting to test now! This is still a work in progress, but I wanted to post what I have so far. I'm currently around Step 8. Most things are checking out!
* When I first visited the audit page, I got this overlay that asks me to Sign In (but I'm already signed in). 
![image](https://user-images.githubusercontent.com/25534091/46578202-15152580-c9ae-11e8-9992-e49b7f0a3bb3.png)
* There's a little bit of funkiness with distances updating. I had a direct jump from 0.1 --> 0.3 --> 0.5 miles travelled. I kept clicking the forward GSV arrow, didn't place any labels/pan around the screen. The distances seem to update once I reach an intersection, rather than in the middle of the street. 
* Not sure if this is expected behavior, but when I click the "Jump" button, my percentage complete sometimes jumps by typically +/- 3%. This happens for both "I want to explore another area!" and "I can't go the direction you want me to walk" options. After I hit it ~6 times on the 1000ft mission, I got a mission complete message. (I was at 35% originally).
* When I refresh the page, distance travelled + number of labels placed seem to stick, but mission completion percentages do not. I was at ~96% for one mission when I refreshed and it took me back to 16%.

misaugstad commented:
Thanks for all this!!

>When I first visited the audit page, I got this overlay that asks me to Sign In (but I'm already signed in).

This is only happening because we don't have record of whether you completed the tutorial any more. If you complete the tutorial as a registered user with the new mission code, you shouldn't see that again.

>There's a little bit of funkiness with distances updating. I had a direct jump from 0.1 --> 0.3 --> 0.5 miles travelled. I kept clicking the forward GSV arrow, didn't place any labels/pan around the screen. The distances seem to update once I reach an intersection, rather than in the middle of the street.

Interesting! I don't remember changing that, so it has probably always been like this, but I might as well fix it now since we are here, thanks! :)

>Not sure if this is expected behavior, but when I click the "Jump" button, my percentage complete sometimes jumps by typically +/- 3%. This happens for both "I want to explore another area!" and "I can't go the direction you want me to walk" options. After I hit it ~6 times on the 1000ft mission, I got a mission complete message. (I was at 35% originally).

For the "I can't go the direction you want me to walk" option, this is expected behavior. If it isn't possible to complete the street, we want to give the user credit for trying. I'll look into the behavior for the "I want to explore another area" option to see if I can make it cleaner. Though this was an existing problem (talked about in issue #315)

>When I refresh the page, distance travelled + number of labels placed seem to stick, but mission completion percentages do not. I was at ~96% for one mission when I refreshed and it took me back to 16%.

That one REALLY surprises me, and I haven't been able to reproduce it. Let me know if you are able to produce that behavior again. And if you can, let me know how consistently you can reproduce it, and any other info you can think of about how you produce the bug (e.g., are you walking or jumping, what is the value in the `distance_progress` column for you mission before and after you refresh the page).

>I also noticed that sbt is printing out stuff. Not sure when these started appearing (wasn't watching closely).

Well you found a print statement I forgot to delete, thanks! And the other thing is some error logging that I added in. I noticed that sometimes timestamps are null for labels for some reason, so I added in that error logging. The bug that causes it was there before this PR though.

Thanks so much for the help and keep it up! I'll work on the distance updating only when finishing a street and the "I want to explore another area" progress bar behavior now!

misaugstad commented:
>There's a little bit of funkiness with distances updating. I had a direct jump from 0.1 --> 0.3 --> 0.5 miles travelled. I kept clicking the forward GSV arrow, didn't place any labels/pan around the screen. The distances seem to update once I reach an intersection, rather than in the middle of the street.

Upon further reflection, I think this should remain as-is. We only consider a street to be audited once it is complete. So if a user audits only part of a street and refreshes the page, they could still be assigned that street to audit later. If we continually increase that number partway through a street, either that number will decrease when a user refreshes the page, OR that number could add up to a distance that is higher than the total distance in that neighborhood in the end. I don't think we want either, and it is best to just have it update less frequently. If we allow users to resume their same audit-task partway through, then we could change when it updates, but for now it is in an optimal state.

misaugstad commented:
And I have now fixed the issue of the mission progress changing when jumping and choosing the "I want to explore another area" option! All of your initial concerns have been addressed!

aileenzeng commented:
> That one REALLY surprises me, and I haven't been able to reproduce it. Let me know if you are able to produce that behavior again. And if you can, let me know how consistently you can reproduce it, and any other info you can think of about how you produce the bug (e.g., are you walking or jumping, what is the value in the distance_progress column for you mission before and after you refresh the page).

Hmm... I haven't been able to reproduce it either. But from what I remember, I was walking (hitting the GSV arrows), and refreshed the page in the middle of the street. 

> Check that there are the correct number of completed missions in the mission table. There should be a max of 1 incomplete mission per user per neighborhood.

It looks like I have duplicate incomplete missions for region 346. Their time stamps are really close together, so maybe something got double-logged? Not sure how to reproduce - this wasn't something I had intentionally tried to do.
![image](https://user-images.githubusercontent.com/25534091/46648229-f5663480-cb48-11e8-9061-60a999b74df4.png)

More stuff: 
* I audited 54% of the neighborhood (1000 meter mission) as a registered user. I jumped around using the "I cannot walk anymore option" ~4 times, and got the Mission Start modal (skipped the mission complete modal). This only happened one time - I'm not sure I can reliably reproduce it, since I've used the jump option a lot, and this is the first time that it's happened.
* This probably isn't too big of a deal, but I noticed that if I went off the audit path and met up back with the route, it'll mark the entire distance in between when I was initially on the path/when I came back onto the path as complete. (EX: taking the small side street on the right around the main audit route).
![image](https://user-images.githubusercontent.com/25534091/46648995-3449b980-cb4c-11e8-8eda-99a5de92053b.png)
* When I'm jumping around to complete missions (as a registered/anonymous user), if I don't complete any street, I won't get any green streets (not sure if this is on purpose or not) on the Mission Complete modal. I'll see more gray streets the mission after though.

Everything else has checked out for registered/anonymous users!

aileenzeng commented:
Just tested the Turker stuff - everything checks out!

misaugstad commented:
>Hmm... I haven't been able to reproduce it either. But from what I remember, I was walking (hitting the GSV arrows), and refreshed the page in the middle of the street.

Mk. We will both keep testing it more thoroughly to make sure that doesn't come up. But that is one of the main things I fixed, so I'm tempted to think that is some JS not being updated in a browser or something :sweat_smile: but definitely needs more testing then.

>It looks like I have duplicate incomplete missions for region 346. Their time stamps are really close together, so maybe something got double-logged? Not sure how to reproduce - this wasn't something I had intentionally tried to do.

Hmm okay I will do more testing to see if I can figure out what is causing that.

>I audited 54% of the neighborhood (1000 meter mission) as a registered user. I jumped around using the "I cannot walk anymore option" ~4 times, and got the Mission Start modal (skipped the mission complete modal). This only happened one time - I'm not sure I can reliably reproduce it, since I've used the jump option a lot, and this is the first time that it's happened.

Most likely what happened is that you had an extra click somewhere on the screen which skipped you through the mission complete screen. It turns out that you don't have to click on the actual "Okay" button, you just have to click anywhere on the screen to advance. So sometimes when I'm jumping repeatedly I accidentally skip through it.

>This probably isn't too big of a deal, but I noticed that if I went off the audit path and met up back with the route, it'll mark the entire distance in between when I was initially on the path/when I came back onto the path as complete. (EX: taking the small side street on the right around the main audit route).

Yeah unfortunately that is how it works right now! :grin: 

>When I'm jumping around to complete missions (as a registered/anonymous user), if I don't complete any street, I won't get any green streets (not sure if this is on purpose or not) on the Mission Complete modal. I'll see more gray streets the mission after though.

Yeah that is a known bug that shouldn't be too hard to fix TBH (#495 ) -- feel free to add any other info you can to that issue thread!

>Everything else has checked out for registered/anonymous users!

Awesome!!

>Just tested the Turker stuff - everything checks out!

Amazing!!!

Okay I will try to figure out why that duplicate mission appeared. From your screenshot, it also looks like the `mission_start` and `mission_end` columns have the same value, which seems wrong :wink: so I'll take a look at that too! Once that is implemented, let's both do one more round of testing where we specifically see if we can reproduce the bug with the mission completion percentage going down and for duplicate missions. Then we should be good to go :fist_oncoming:

misaugstad commented:
Okay I moved some code around and added a synchronized code block to deal with the race conditions that I believe were the root cause of the duplicate missions. I also made it so that the `mission_end` column updates continuously while the user is auditing.

Alright @aileenzeng last round of testing, let's do this! Thanks so much for the help with testing!

aileenzeng commented:
I haven't been able to recreate the mission start time bug or percentage changing bug!

When I hit the "skip tutorial" button, the onboarding is marked as complete in the `mission` table. Is there a way that we differentiate between users that completed the tutorial / skipped the tutorial?

misaugstad commented:
>When I hit the "skip tutorial" button, the onboarding is marked as complete in the mission table. Is there a way that we differentiate between users that completed the tutorial / skipped the tutorial?

No I don't believe so... I think the best way to deal with this might be to log that they skipped the tutorial either in the `webpage_activity` table or the `audit_task_interaction` table, what do you think? Or do you think it would be more appropriate to add a `skipped` column to the `mission` table?

jonfroehlich commented:
Definitely think it should logged. I lean towards logging critical things
in their own table... but don't feel strongly about this one.

On Tue, Oct 16, 2018 at 1:47 PM Mikey Saugstad <notifications@github.com>
wrote:

> When I hit the "skip tutorial" button, the onboarding is marked as
> complete in the mission table. Is there a way that we differentiate between
> users that completed the tutorial / skipped the tutorial?
>
> No I don't believe so... I think the best way to deal with this might be
> to log that they skipped the tutorial either in the webpage_activity
> table or the audit_task_interaction table, what do you think? Or do you
> think it would be more appropriate to add a skipped column to the mission
> table?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProjectSidewalk/SidewalkWebpage/pull/1319#issuecomment-430394368>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABi-9a1nQ4lYeRgmwWBe0r_eYNQDJNxDks5ulkXAgaJpZM4W9om9>
> .
>


-- 
Jon Froehlich
Associate Professor
Paul G. Allen School of Computer Science & Engineering
University of Washington
http://www.cs.umd.edu/~jonf/
http://makeabilitylab.io
@jonfroehlich <https://twitter.com/jonfroehlich> - Twitter

misaugstad commented:
I think it will rarely be used, but it will be make analysis a lot easier if it is added as a column in the table, so that's what I'm going to do.

aileenzeng commented:
Oops I forgot to respond yesterday, but I definitely like adding a column!

jonfroehlich commented:
Yep. I like that. Also, columns maybe to mark labels as tutorial labels,
etc. would likely make things easier too--so I'm generally in favor of this
approach. Thoughts?

On Wed, Oct 17, 2018 at 9:57 AM Mikey Saugstad <notifications@github.com>
wrote:

> I think it will rarely be used, but it will be make analysis a lot easier
> if it is added as a column in the table, so that's what I'm going to do.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProjectSidewalk/SidewalkWebpage/pull/1319#issuecomment-430706982>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABi-9RZiJfj6RxQpO2kYn82B_M7SVv9Hks5ul2FugaJpZM4W9om9>
> .
>


-- 
Jon Froehlich
Associate Professor
Paul G. Allen School of Computer Science & Engineering
University of Washington
http://www.cs.umd.edu/~jonf/
http://makeabilitylab.io
@jonfroehlich <https://twitter.com/jonfroehlich> - Twitter

misaugstad commented:
>Also, columns maybe to mark labels as tutorial labels,
etc. would likely make things easier too

Yep and that is actually a particularly good idea!

misaugstad commented:
@aileenzeng Now when skipping the tutorial, you should see the action "Onboarding_Skip" logged in the `audit_task_interaction` table right before the "Onboarding_End" action. Also, there is a new `skipped` column in the `mission` table. This should only be marked `true` if the user skips the tutorial, and in no other situation. Try doing the tutorial all the way through, and auditing a couple missions to make sure that the `skipped` column is not erroneously marked true.

aileenzeng commented:
It checks out! Everything works nicely!

misaugstad commented:
Amazing!! Time for the merge :sunglasses:
